feat(install): 1. and 3. phase from design docs

jkuri opened the pull request:
Install example: `ng install moment --typings moment-node`
Uninstall examle: `ng uninstall lodash --typings lodash`

cironunes commented:
no test? :)

jkuri commented:
Not yet :-)

hansl commented on addon/ng2/tasks/uninstall.js:
```diff
@@ -0,0 +1,157 @@
+  uninstallProcedure: function() {
+    const that = this;
+
+    return that.removeFromPackageJSON(that.package)
```
The goal of promises was to avoid this kind of thing :)  you can simplify this like this:

```typescript
    return that.removeFromPackageJSON(that.package)
      .then(() => that.npmUninstall(that.package))
      .then(() => that.unlinkPackage(that.package))
      .then(() => that.removeFromSystemJSConfig(that.package))
      .then(() => {
        that.ui.stopProgress();
        if (that.typings) {
          that.ui.startProgress(chalk.green('Uninstalling typings:', chalk.yellow(that.typings)), chalk.green('.'));
          return that.typingsUninstall(that.typings)
        }
      })
      .then(() => {
        that.ui.stopProgress();
        that.announceOKCompletion();
      });
```

jkuri replied:
thanks Ciro.

hansl commented on addon/ng2/tasks/uninstall.js:
```diff
@@ -0,0 +1,157 @@
+              });
+            }
+            else {
+              that.announceOKCompletion();
```
I want to get away from this and have a thirdparty for this. I think it's fine not to have an announcer right now.

hansl commented on addon/ng2/tasks/uninstall.js:
```diff
@@ -0,0 +1,157 @@
+  },
+
+  npmUninstall: function(pkg) {
+    let cmd = ['npm uninstall'];
```
Could you use the ember-cli tasks for this? It will be cleaner. As an example, look at how we call `npm install` in the `init` command.

jkuri replied:
`ember-cli` removed the `npm` tasks, please see https://github.com/ember-cli/ember-cli/blob/master/lib/commands/install-npm.js

hansl replied:
That's the command, not the task.

https://github.com/ember-cli/ember-cli/blob/master/lib/tasks/npm-install.js

hansl commented on addon/ng2/tasks/uninstall.js:
```diff
@@ -0,0 +1,157 @@
+  },
+
+  typingsUninstall: function(pkg) {
+    let cmd = ['typings uninstall'];
```
Please create new tasks for this. It might be something we want to do in other commands.

jkuri replied:
Sure.

hansl commented on addon/ng2/tasks/uninstall.js:
```diff
@@ -0,0 +1,157 @@
+  },
+
+  removeFromSystemJSConfig: function(pkg) {
+    const systemPath = path.resolve(process.cwd(), 'src', 'config', 'system.config.js');
```
Doesn't this only work for modules that have a proper SystemJS bundle?

jkuri replied:
No, it works for plain `js` files... I added separate SystemJS config as you stated in the docs.

hansl commented on addon/ng2/tasks/install.js:
```diff
@@ -0,0 +1,178 @@
+    const that = this;
+
+    return that.npmInstall(that.package)
+    .then(() => {
```
See below.

hansl commented on addon/ng2/tasks/install.js:
```diff
@@ -0,0 +1,178 @@
+/* jshint node: true, esnext: true */
```
As a general rule, tasks should perform a single task and be composed of other tasks. Those other tasks are reusable.

Ask yourself how you can define the `install` task into steps, and those steps should become their own tasks. Stuff like installing typings should be its own task.

jkuri replied:
@hansl I would need some of your help here. 
There are/will be actually more possible procedures, for example just `ng install` to install all the packages in the config file. The install procedure can go in many ways according to the result the previous task responses. I am not sure how to achieve that with promises that I have now and I don't want to repeat calling same task multiple times. 
Hope this description of the problem I am trying to avoid is clear.

hansl commented on addon/ng2/tasks/install.js:
```diff
@@ -0,0 +1,178 @@
+    return exec(cmd.concat(options).join(' '));
+  },
+
+  storeInPackageJSON: function(pkg) {
```
That seems like it's just adding `angular-cli: { name: 'package' }` to the package.json. There was a discussion about moving that information to a separate file as it will become really huge really fast. I think it could be useful to have another file. What do you think?

jkuri replied:
I would prefer single config file, but can make a separate one if you think this should be more transparent. Otherwise is not a big deal having a large JSON file the users should not update it directly anyway.

hansl replied:
We will need a separate configuration file, actually. It will get a lot of stuff in it.

hansl commented on addon/ng2/tasks/install.js:
```diff
@@ -0,0 +1,178 @@
+  },
+
+  copyPackages: function() {
+    const packages = require(packageJSON)['angular-cli'].packages || [];
```
I'm confused as to what this does.

hansl commented on addon/ng2/tasks/uninstall.js:
```diff
@@ -0,0 +1,157 @@
+    this.package = options.package;
+    this.typings = options.typings;
+
+    this.disableLogger();
```
What are you trying to achieve here? If you want to intercept child processes' stdout/stderr you should use `spawn` instead of `exec`.

jkuri replied:
This is indeed needed, I think. `npm install` otherwise outputs some info about the process that we don't want. I actually took that from one existing `ember-cli` task, but not sure which.

hansl commented on addon/ng2/tasks/uninstall.js:
```diff
@@ -0,0 +1,157 @@
+  },
+
+  disableLogger: function() {
+    this.oldLog = console.log;
```
This is dangerous for many reasons.

hansl commented on addon/ng2/tasks/uninstall.js:
```diff
@@ -0,0 +1,93 @@
+    this.ui.writeLine(chalk.red(this.completionErrorMessage));
+  },
+
+  existsSync: function(path) {
```
This seems unused.

hansl commented on addon/ng2/tasks/install.js:
```diff
@@ -0,0 +1,81 @@
+    this.ui.writeLine(chalk.red(this.completionErrorMessage));
+  },
+
+  existsSync: function(path) {
```
Unused.

hansl commented on addon/ng2/blueprints/ng2/files/src/config/system.config.js:
```diff
@@ -0,0 +1,8 @@
+var systemConfig = {
```
Could you use `System.config(...)` here directly?

jkuri replied:
Sure, its nicer. Done.

hansl commented on addon/ng2/tasks/npm-install.js:
```diff
@@ -0,0 +1,17 @@
+/* jshint node: true, esversion: 6 */
```
What's different between this and the `npm-install` task from ember-cli?

jkuri replied:
`ember-cli` task uses `console.log` interceptor which is dangerous.

hansl replied:
I think in this case it's okay, since it's not our code. Unless you see a problem caused by their code.

jkuri replied:
I also don't like that part: https://github.com/ember-cli/ember-cli/blob/master/lib/tasks/npm-task.js#L18. It causes problems on Windows, there were a lot of people complaining about it, the fix was to `npm link npm`.

filipesilva replied:
`npm link npm` caused a lot of problems with testing. See https://github.com/angular/angular-cli/pull/198

hansl replied:
@jkuri I'm okay with the task then.  For documentation, could you point to the issue on ember-cli about problems on Windows. Ideally add a link to it in a comment in the file describing why we do our own task.

jkuri replied:
Posted here: https://github.com/ember-cli/ember-cli/issues/5637

hansl replied:
Stefan answered that issue with a request to follow their template.

Also, where did you find out about that bug? Did you test it yourself? Could you try with ember-cli (without angular) to make sure it's not a bug on our side?

jkuri replied:
Yes, I saw, I will prepare the demo which is requested in the issue.

I am pretty sure it is not the issue on our side. The initial `ng install` used that task and the were Windows users complaining about it (including @filipesilva). I will also test it myself on the VM as you suggested.

hansl commented on addon/ng2/tasks/install.js:
```diff
@@ -0,0 +1,72 @@
+    const systemPath = path.resolve(process.cwd(), 'src', 'config', 'system.config.js');
+    let systemContents = fs.readFileSync(systemPath, 'utf8');
+
+    systemContents = systemContents.split('\n');
```
You can simplify this with something like this:

```javascript
const systemContents = fs.readFileSync(systemPath, 'utf8');

const jsonContent = systemContents.match(/^[^\{]((?:.|\n)+)\);$/)[1];
const json = JSON.parse(jsonContent);
```

jkuri replied:
Thanks for this regex.

hansl commented on addon/ng2/tasks/uninstall.js:
```diff
@@ -0,0 +1,83 @@
+    const systemPath = path.resolve(process.cwd(), 'src', 'config', 'system.config.js');
+    let systemContents = fs.readFileSync(systemPath, 'utf8');
+
+    systemContents = systemContents.split('\n');
```
See above. Please move both functions into a separate module and require that module.

Something like:

```typescript
export function loadSystemJson(systemPath: string): { [name: string]: any } {
  ...
}
export function saveSystemJson(systemPath: string, json: { [name: string]: any }) {
  ...
}```

gkalpak commented on addon/ng2/utilities/systemjs-helper.ts:
```diff
@@ -0,0 +1,15 @@
+
+export function loadSystemJson(systemPath: string): { [name: string]: any } {
+  const systemContents: any = fs.readFileSync(systemPath, 'utf8');
+  const jsonContent: any = systemContents.match(/^[^\{]*((?:.|\n)+)\);.*?$/m)[1];
```
WRT the RegExp:

* The `?` at the end (before the `$`) seems redundant.

* `(?:.|\n)` will not match _any_ character (as seems to be the intention). For example, it will not match `\r`. So, users on Windows (where the newline sequence is `\r\n`) will run into errors.  
    If you want to match anything (including newline characters), then `[\s\S]` is the way to go (in JS at least).

hansl replied:
Sounds better. Thanks for keeping an eye our for us!

jkuri replied:
Thanks @gkalpak, updated.

hansl commented:
Okay please add at least one e2e test for this as this is a sensitive command. We really want to prevent regression.

Try something like:
```
ng new
ng install moment
ng build
```

And verify the build is what you expect, or try to `ng serve` and access the page. Whichever you think it's easier.

jkuri commented:
Sure, I agree for tests. But this now actually has no meaning because packages are not copied, this is basically just `npm install` and `typings install` for now, nothing is automated. Would like to chat about that with you.
I will prepare the tests anyway for later.

hansl commented:
Please fix the CI errors.

jkuri commented:
As we agreed with @hansl I implemented the css preprocessors support for `sass/less/stylus/compass`. 
This works right off after installing for example `ng install sass` which installs sass into project directory and it's then used (with some hacks) in the `angular-cli` side. This is the best option we have for now. 
I also added the e2e tests where I actually write some styles and checks if they are compiled correctly.

jkuri commented:
CI fixed. Its working on all platforms now.

zackarychapple commented:
Wow @jkuri this is going to be nice to have.  Good work :+1:

hansl commented on addon/ng2/commands/install.js:
```diff
@@ -0,0 +1,39 @@
+  works: 'insideProject',
+
+  availableOptions: [
+    { name: 'typings', type: String, aliases: ['t'], description: 'Installs specified typings' }
```
I'm not sure we need this. We should check if a typings exist for a node package we're trying to install. Is there a case where this doesn't work?

jkuri replied:
Yes, for example (and seems like most common one) `moment` has `moment-node` typings. I believe there's a lot of typings that name mismatch the original node package name.

hansl commented on lib/broccoli/angular2-app.js:
```diff
@@ -87,13 +99,80 @@ Angular2App.prototype.toTree = function () {
+  ];
+
+  try {
+    if (process.platform === 'win32') {
```
That's way too much code. Can't you just ```
// 
trees = trees.concat(
  require('./angular-broccoli-sass').makeBroccoliTree(sourceDir),
  require('./angular-broccoli-less').makeBroccoliTree(sourceDir),
  require('./angular-broccoli-compass').makeBroccoliTree(sourceDir)
]).filter(x => !!x);
```

etc etc.  The `broccoli-sass` script should make the Funnel and check if sass is installed. If not, `return null`.

What do you think?

jkuri replied:
Yes, you are right, it will be a lot cleaner.

jkuri replied:
Fixed according to your tips.

hansl commented on addon/ng2/blueprints/ng2/files/package.json:
```diff
@@ -5,7 +5,7 @@
   "scripts": {
     "start": "ng server",
-    "postinstall": "typings install --ambient",
+    "postinstall": "typings install",
```
Are you sure you want to ignore ambient typings?

jkuri replied:
Fixed.

hansl commented on .eslintignore:
```diff
@@ -0,0 +1 @@
+addon/ng2/blueprints/ng2/files/src/config/system.config.js
```
Why do we want to ignore that single file? It seems like it should be linted.

jkuri replied:
Because of the double quotes which are mandatory for `JSON.parse` in `systemjs-helper.ts`.

gkalpak replied:
Maybe using `eval` instead of `JSON.parse` (in `loadSystemJson()`) would be a viable alternative and remove the need for double quotes.

I don't think there are any performance or security concerns here, but still `eval` is ugly :confused:

hansl commented on addon/ng2/tasks/install.ts:
```diff
@@ -0,0 +1,128 @@
+    return new Promise(resolve => {
+      let promises = [];
+
+      that.packages.forEach(p => {
```
This sounds more like a `map()` than a `forEach()`.

hansl commented on lib/broccoli/angular-broccoli-less.js:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    build() {
+      const that = this;
```
You don't need this since you're using `=>` everywhere.

jkuri commented on addon/ng2/tasks/install.ts:
```diff
@@ -0,0 +1,128 @@
+  completionErrorMessage: 'Error installing package.',
+
+  run: function(options) {
+    this.packages = options.packages;
```
It gets `object` with data from `install` command, I just name it `options` in case we'll add something later, currently is just the packages it gets from arguments.

hansl commented on addon/ng2/tasks/uninstall.ts:
```diff
@@ -0,0 +1,84 @@
+    this.packages = this.packages.filter(p => !/node-sass|stylus|less|compass-importer/.test(p));
+    let typingsList = [];
+
+    return new Promise(resolve => {
```
You don't need that, just `return Promise.all(...).then(...)`. `then()` returns a Promise.

jkuri replied:
Thanks, updated.

hansl commented on lib/broccoli/angular-broccoli-less.js:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    compile(fileName, inputPath, outputPath) {
+      return new Promise(resolve => {
```
You can do something like this to simplify:

```javascript
    compile(fileName, inputPath, outputPath) {
        let content = fs.readFileSync(fileName, 'utf8');

        return less.render(content)
        .then(output => {
          let filePath = fileName.replace(inputPath, outputPath).replace(/\.less$/, '.css');
          fse.outputFileSync(filePath, output.css, 'utf8');
        });
    }
```

jkuri replied:
Fixed.

jkuri commented:
Closing this. If nothing else, I learned something from this PR :-) Thanks @hansl!

choeller commented:
@jkuri Do I understand it correct, that this did not make it into the master? And if so - why?! :flushed: I think this a super essential feature and many people are looking forward to it...

jkuri commented:
We had some problems with typings. This will be implemented but cannot tell you when because I don't know it myself :) If you are interested for sass/less support you can follow it [here](https://github.com/angular/angular-cli/pull/388), the rest of the things from design docs will follow later.

choeller commented:
Oh, ok - that's sad to hear :/ so to use 3rd party libraries like ngrx we still need to copy the sources to the project?

jkuri commented:
Unfortunately yes

hansl commented:
I just want to clarify that you don't need to copy files to your project
for ngrx. You should simply be able to `npm install ngrx` and add those
files to your venderNpmFiles.

We are working on the production build story to use CDN versions of those
vendor files.

On Mon, Apr 4, 2016 at 10:41 AM, Jan Kuri <notifications@github.com> wrote:

> Unfortunately yes
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/angular/angular-cli/pull/294#issuecomment-205410353>
>
